### ISSUE TITLE
히든 레이어에서 가중치

### ISSUE THREAD

**Issue by thanksto** [codingeverybody/codingyahac#1075]
# 해결하고자 하는 문제
히든레이어를 쌓다보면 학습이 잘되는 것도 알겠고 학습을 위해 히든레이어나 퍼셉트론의 수가
어떻게 되어야 하는 질문은 다른분이 올린 질문의 답변을 통해 알게 되었습니다.
그런데 하나 궁금한게 히든레이어를 쌓아 model을 만들었다고 했을 때
독립변수에 따라 종속변수가 어떻게 바뀌는지에 대해서는 가중치를 보는 방법이 없을까요?
강의를 보고 가중치를 구하는 법으로 model.get_weight을 했을 때는
독립변수가 히든레이어를 구성하는 퍼셉트론의 가중치만 나와서 
히든레이어가 있을경우 입력값에 따른 출력값 가중치는 어떻게 보는지 궁금해서 질문드립니다.

# 코드 혹은 오류
```
   여기에 코드를 적어주세요. 
```봉수골 개발자 이선비 선생님 tensorflow(python)강의에서 나온대로 앞부분을 코딩했고 
히든레이어 설정한것과 나온값들만 적겠습니다.
코딩한 부분에 문제가 있거나 의문이 있어 올리는 질문은 아니고 다른 궁금점이 있어 질문을 하는것이니
아래 코딩은 다 읽지않으셔도 될거같습니다.


X = tf.keras.layers.Input(shape=[4])
H = tf.keras.layers.Dense(8, activation="swish")(X)
H = tf.keras.layers.Dense(8, activation="swish")(H)
H = tf.keras.layers.Dense(8, activation="swish")(H)
Y = tf.keras.layers.Dense(3, activation='softmax')(H)
model = tf.keras.models.Model(X, Y)
model.compile(loss='categorical_crossentropy',
              metrics='accuracy')

model.get_weights()

[array([[ 0.00510967,  0.18849371,  0.23519738,  0.06812343,  0.3957737 ,
          0.7653156 ,  0.7771267 ,  0.26724827],
        [-0.5740305 ,  0.20782383, -0.2388071 , -0.36593837, -0.53794026,
          0.9569792 ,  0.68816733,  0.47062132],
        [ 0.869755  , -0.31789535,  0.6042885 ,  0.06516431,  0.0303264 ,
         -0.3638371 , -0.26000032, -0.7178851 ],
        [ 1.0827293 ,  0.89222354, -0.04198024,  1.3408904 ,  0.22888507,
         -0.38751715, -1.0340394 ,  0.30636972]], dtype=float32),
 array([-0.4961006 , -0.52114475, -0.43291378, -0.05077836,  0.5951885 ,
         0.5826891 ,  0.5761337 ,  0.405556  ], dtype=float32),
 array([[ 0.12642527, -0.03040598,  0.49148786,  0.5963037 ,  0.31564486,
         -1.0065322 ,  0.62555975, -1.0584615 ],
        [-0.21480921, -0.6540953 , -1.1486428 , -1.0130296 , -0.9204345 ,
         -0.37683308,  0.79819274, -0.5687575 ],
        [-0.14716603, -0.08332343,  0.46304455, -0.17406245,  0.5495715 ,
         -0.21392067,  0.76888084, -0.00729801],
        [-1.5536546 ,  0.5473992 ,  0.65685374, -0.9282774 , -1.0859791 ,
         -0.65848327,  0.35650623,  0.02706131],
        [-0.7858825 ,  1.3193911 ,  1.702033  ,  0.6800798 ,  0.1576581 ,
         -0.3106156 , -0.36482632,  0.06264346],
        [ 0.27443764, -0.70298886, -1.0632628 ,  0.03295332, -0.18896589,
          0.7411317 , -0.1786045 ,  0.33218452],
        [ 0.5800038 , -0.2460246 , -0.6995589 , -0.16402411,  0.22878571,
          0.81428224, -0.26576072,  0.06069661],
        [ 0.08188698,  0.12110486, -1.7168068 , -1.1357162 , -2.5191617 ,
          1.0983624 , -0.47592613, -1.4685185 ]], dtype=float32),
 array([ 0.22052866,  0.33651564,  0.24834037,  0.24334091,  0.18225437,
         0.5343299 , -0.18154098,  0.501662  ], dtype=float32),
 array([[-0.16645956,  0.26653504, -2.934371  ,  2.6391842 ,  2.2458708 ,
         -2.222622  ,  1.4588435 ,  0.5716353 ],
        [-1.8451242 , -1.3158338 , -0.11176367,  2.6099918 , -0.55108666,
         -1.0927507 ,  0.970209  ,  1.9482975 ],
        [-1.5239317 , -0.7352833 ,  1.3815743 ,  3.176794  ,  0.50711066,
         -1.2172371 ,  0.27851582,  1.4902238 ],
        [ 0.8035957 ,  0.6766161 , -1.4237034 ,  2.046594  , -0.9415446 ,
         -2.1842859 , -2.2096179 , -0.1003596 ],
        [ 0.70723015,  0.40857673, -1.3599708 ,  3.6405644 , -3.2972229 ,
         -3.828363  , -1.4715706 , -1.3661366 ],
        [ 0.7106915 ,  0.9910289 , -2.1139772 ,  1.9711637 ,  3.1183875 ,
         -0.6623332 , -0.0564611 ,  2.1513765 ],
        [-0.30924052,  0.20429453,  1.36848   , -0.04908543, -1.9227031 ,
          0.48405862,  1.1547936 , -0.9868787 ],
        [-2.6143563 , -2.3288016 ,  2.5268424 ,  1.301177  ,  0.8488148 ,
          1.2926857 ,  2.740687  , -0.72256345]], dtype=float32),
 array([ 1.0586742 ,  0.56812644,  0.12430122, -2.102996  ,  0.357384  ,
         0.78278506, -0.31953523, -1.1065624 ], dtype=float32),
 array([[-3.1162407e+00,  1.6012765e+00, -2.0056293e+00],
        [-3.3999872e+00,  1.2279552e+00, -9.1597730e-01],
        [ 1.0380577e+00, -5.0067806e+00,  4.9771390e+00],
        [ 3.1453401e-01, -4.8053970e+00,  3.9660199e+00],
        [-1.8295540e-03,  4.3811479e+00, -4.1361694e+00],
        [ 2.0069788e+00, -2.4459131e+00,  2.2198794e+00],
        [-2.3558977e+00, -3.4577177e+00,  2.3430188e+00],
        [ 6.4071264e+00, -2.1484762e-01, -1.7795980e+00]], dtype=float32),
 array([-10.585511  ,   0.8268408 ,  -0.31131753], dtype=float32)]


# 환경
사용중인 운영체제, 언어, 라이브러리의 버전을 적어주세요.
python(tensorflow)

**Comment by JeongsangBaek** [comment]
저는 모델 예측이 내가 생각하는대로 동작하는지, 그 예측값에 가장 많은 영향을 주는 퍼셉트론과 그 퍼셉트론의 weight이 어떻게 설정되어 있는지, 원하는 값이 나오지 않았다면 어떤 퍼셉트론을 '조작'하면 원하는 방향으로 동작하는지에 대한 확인을 도와주는 툴이 있는지에 대한 질문으로 이해했습니다. 제가 이해한 게 맞다면 그러한 처리를 도와주는 [What-If Tool](https://pair-code.github.io/what-if-tool/)이 있습니다.

**Comment by RayleighKim** [comment]
사실, 질문을 다시 한 번 정리해주시기를 부탁드립니다.

제가 질문을 이해한 것이 올바르다면..

구성하신 네트워크는 인풋레이어가 아웃풋레이어에 직접 연결되어 있는 형태가 아닙니다.
따라서, 가장 단순한 선형모델 처럼 인풋레이어와 아웃풋레이어에 직접 연결된 가중치는 존재하지 않습니다.

예를 들자면, 
sepal length의 크기가 커진다면 setosa일 확률이 커지더라! <-- 단순한 선형모델에서는 가능합니다. 말씀하신 가중치를 관찰하면 됩니다. 하지만 구성하신 네트워크에서는 존재하지 않는 가중치입니다.

**Comment by thanksto** [comment]
아 두분다 답변해주셔서 감사합니다. 제가 확인을 늦게했네요
제가 한 질문은 RayleighKim님이 말해주신 답변내용입니다. 질문을 좀 애매하게 했나보네요 그래도 다른거의 해답도 들을 수 있어 감사합니다.
하나 더 궁금한게 RayleighKim님 말씀대로 hidden laryer가 존재하면 input layer가 hidden layer에 연결되다 보니 input layer와 output layer가 직접 연결되는게 아니다보니 직접 연결된 가중치가 존재하지 않는다고 하셨는데 그러면 결국에는 output값에 대해서 어떤 input이 중요한지 알 방법이 없다는건가요?

**Comment by thanksto** [comment]
> 사실, 질문을 다시 한 번 정리해주시기를 부탁드립니다.
> 
> 제가 질문을 이해한 것이 올바르다면..
> 
> 구성하신 네트워크는 인풋레이어가 아웃풋레이어에 직접 연결되어 있는 형태가 아닙니다.
> 따라서, 가장 단순한 선형모델 처럼 인풋레이어와 아웃풋레이어에 직접 연결된 가중치는 존재하지 않습니다.
> 
> 예를 들자면,
> sepal length의 크기가 커진다면 setosa일 확률이 커지더라! <-- 단순한 선형모델에서는 가능합니다. 말씀하신 가중치를 관찰하면 됩니다. 하지만 구성하신 네트워크에서는 존재하지 않는 가중치입니다.


아 두분다 답변해주셔서 감사합니다. 제가 확인을 늦게했네요
제가 한 질문은 RayleighKim님이 말해주신 답변내용입니다. 질문을 좀 애매하게 했나보네요 그래도 다른거의 해답도 들을 수 있어 감사합니다.
하나 더 궁금한게 RayleighKim님 말씀대로 hidden laryer가 존재하면 input layer가 hidden layer에 연결되다 보니 input layer와 output layer가 직접 연결되는게 아니다보니 직접 연결된 가중치가 존재하지 않는다고 하셨는데 그러면 결국에는 output값에 대해서 어떤 input이 중요한지 알 방법이 없다는건가요?

**Comment by RayleighKim** [comment]
방법들이 없는 것은 아니지만, 간결하게 느껴지진 않습니다.
Permutation Importance 같은 방법들이 있습니다. 그러나 이는 가중치를 이용한 해석은 아닙니다.
혹은 최근에 설명가능한 인공지능이라 하여,, 국소적인 부분에서 설명을 시도하는 테크닉도 있긴 합니다. (LIME)

하지만, 딥러닝을 더 공부하신다면..
이리 저리 멋진 설명들을 해낼 테크닉들이 존재합니다.

**Comment by thanksto** [comment]
답변감사합니다!
